src/snowflake/snowpark/_internal/utils.py

```diff
@@ -696,6 +696,71 @@ def warning(name: str, text: str, warning_times: int = 1) -> None:
     warning_dict[name].warning(text)
 
 
+def publicapi(func) -> Callable:
+    """decorator to safeguard public APIs with global feature flags."""
+
+    @functools.wraps(func)
+    def func_call_wrapper(*args, **kwargs):
+        # warning(func.__qualname__, warning_text)
+
+        # Handle AST encoding, by modifying default behavior.
+        # If a function supports AST encoding, it must have a parameter _emit_ast.
+        # If now _emit_ast is passed as part of kwargs (we do not allow for the positional syntax!)
+        # then we use this value directly. If not, but the function supports _emit_ast,
+        # we override _emit_ast with the session parameter.
+        if "_emit_ast" in func.__code__.co_varnames and "_emit_ast" not in kwargs:
```

Comment:
I agree, we could rewrite/condense this function a lot by having the Snowpark API adhere to a better structure. For now as you can see it is quite long to support the existing use cases. I would actually suggest to decouple this PR from an optimization effort. I.e., it would be very helpful to have a followup that optimizes this decorator and simplifies it together with the Snowpark structure. This would untangle things and make it easier to understand for readers what changes we did (e.g., performance related changes). Wdyt?

Reply:
This is all temporary code anyway, so probably not worth worrying about too much.

Reply:
Marking this as resolved for now, we can improve the code in subsequent PRs if need be.